zenoh-flow/src/io/input.rs

```diff
@@ -0,0 +1,190 @@
+// implemented on it: `keys()` for one.
+impl Deref for Inputs {
+    type Target = HashMap<PortId, Vec<flume::Receiver<LinkMessage>>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.hmap
+    }
+}
+
+impl Inputs {
+    pub(crate) fn new() -> Self {
+        Self {
+            hmap: HashMap::default(),
+        }
+    }
+
+    /// Insert the `flume::Receiver` in the [`Inputs`](`Inputs`), creating the entry if needed in
+    /// the internal `HashMap`.
+    pub(crate) fn insert(&mut self, port_id: PortId, rx: flume::Receiver<LinkMessage>) {
+        self.hmap
+            .entry(port_id)
+            .or_insert_with(Vec::default)
+            .push(rx)
+    }
+
+    /// Returns the typed [`Input<T>`](`Input`) associated to the provided `port_id`, if one is
+    /// associated, otherwise `None` is returned.
+    ///
+    /// ## Performance
+    ///
+    /// With a typed [`Input<T>`](`Input`), Zenoh-Flow will perform operations on the underlying
+    /// data: if the data are received serialized then Zenoh-Flow will deserialize them, if they are
+    /// received "typed" then Zenoh-Flow will check that the type matches what is expected.
+    ///
+    /// If the underlying data is not relevant then these additional operations can be avoided by
+    /// calling `take_raw` and using an [`InputRaw`](`InputRaw`) instead.
+    pub fn take<T: ZFData>(&mut self, port_id: impl AsRef<str>) -> Option<Input<T>> {
+        self.hmap.remove(port_id.as_ref()).map(|receivers| Input {
+            _phantom: PhantomData,
+            input_raw: InputRaw {
+                port_id: port_id.as_ref().into(),
+                receivers,
+            },
+        })
+    }
+
+    /// Returns the [`InputRaw`](`InputRaw`) associated to the provided `port_id`, if one is
+    /// associated, otherwise `None` is returned.
+    ///
+    /// ## Convenience
+    ///
+    /// With an [`InputRaw`](`InputRaw`), Zenoh-Flow will not manipulate the underlying data, for
+    /// instance trying to deserialize them to a certain `T`.
+    ///
+    /// It is thus up to the user to call `try_get::<T>()` on the [`Payload`](`Payload`) and handle
+    /// the error that could surface.
+    ///
+    /// If all the data must be "converted" to the same type `T` then calling `take::<T>()` and
+    /// using a typed [`Input`](`Input`) will be more convenient and as efficient.
+    pub fn take_raw(&mut self, port_id: impl AsRef<str>) -> Option<InputRaw> {
+        self.hmap
+            .remove(port_id.as_ref())
+            .map(|receivers| InputRaw {
+                port_id: port_id.as_ref().into(),
+                receivers,
+            })
+    }
+}
+
+/// An [`InputRaw`](`InputRaw`) exposes the [`LinkMessage`](`LinkMessage`) it receives.
+///
+/// It's primary purpose is to ensure "optimal" performance. This can be useful to implement
+/// behaviour where actual access to the underlying data is irrelevant.
+#[derive(Clone, Debug)]
+pub struct InputRaw {
```

Comment:
What happens if I use an `Output<T>` to send data to an `InputRaw`? 
My guess is that data need to be serialized by the `InputRaw::recv_async`, if not I guess that something bad will happen.

Reply:
Nothing bad happens! :-)

An `Output<T>` will send a `Payload::Typed` to downstream nodes. Zenoh-Flow will thus serialize it (as `T` implements `ZFData`) only if need be.